### PR TITLE
[GEOS-10882] Add an option to remove trailing slash match in OGC APIs

### PR DESCRIPTION
**Pull request by turingtestfail**
[GEOS-10882](https://osgeo-org.atlassian.net/browse/GEOS-10882)

cleanup

testing

global settings test

documentation

moved in page

documentation and move to others

cleanup

remove dispatcher test

included restart note

restconfig test fixes

is online issue

undid changes

test fixes

more tests

more test cleanup

more test cleanup

cleanup

cleanup

  

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geoserver.org/latest/en/developer/policies/committing.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [ ] All the build checks are green ([see automated QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html)).

For core and extension modules:

- [x] New unit tests have been added covering the changes.
- [x] [Documentation](https://github.com/geoserver/geoserver/tree/main/doc/en/user/source) has been updated (if change is visible to end users).
- [ ] The [REST API docs](https://github.com/geoserver/geoserver/tree/main/doc/en/api/1.0.0) have been updated (when changing configuration objects or the REST controllers).
- [x] There is an issue in the [GeoServer Jira](https://osgeo-org.atlassian.net/browse/GEOS/summary) (except for changes that do not affect administrators or end users in any way).
- [x] Commit message(s) must be in the form ``[GEOS-XYZWV] Title of the Jira ticket``.
- [x] Bug fixes and small new features are presented as a single commit.
- [x] Each commit has a single objective (if there are multiple commits, each has a separate JIRA ticket describing its goal).

### CODE REVIEW

File: src/main/src/main/java/org/geoserver/config/GeoServerInfo.java [hunk 1]
```diff
@@ -120,6 +120,10 @@ public interface GeoServerInfo extends Info {
      */
     Boolean isXmlExternalEntitiesEnabled();
 
+    void setTrailingSlashMatch(Boolean trailingSlashMatch);
+
+    Boolean isTrailingSlashMatch();
+
```
Comment (aaime) [inline src/main/src/main/java/org/geoserver/config/GeoServerInfo.java:126]: Javadocs missing

File: src/main/src/main/java/org/geoserver/config/impl/GeoServerInfoImpl.java [hunk 2]
```diff
@@ -50,6 +50,8 @@ public class GeoServerInfoImpl implements GeoServerInfo {
 
     protected Boolean xmlExternalEntitiesEnabled = Boolean.FALSE;
 
+    protected Boolean trailingSlashMatch = Boolean.FALSE;
```
Comment (aaime) [inline src/main/src/main/java/org/geoserver/config/impl/GeoServerInfoImpl.java:53]: Should be true by default, we might decide to set it to false down the line, in preparation for a Spring 6 upgrade (where doing a trailing slash match will require more than just configuration)

File: src/rest/src/main/java/org/geoserver/rest/RestConfiguration.java [hunk 3]
```diff
@@ -208,6 +211,13 @@ public void configurePathMatch(PathMatchConfigurer configurer) {
         helper.setAlwaysUseFullPath(true);
         configurer.setUrlPathHelper(helper);
         configurer.setUseSuffixPatternMatch(true);
+        if (geoServer != null
+                && geoServer.getGlobal() != null
+                && geoServer.getGlobal().isTrailingSlashMatch() != null) {
+            configurer.setUseTrailingSlashMatch(geoServer.getGlobal().isTrailingSlashMatch());
+        } else {
+            configurer.setUseTrailingSlashMatch(true);
+        }
```
Comment (aaime) [inline src/rest/src/main/java/org/geoserver/rest/RestConfiguration.java:220]:
```suggestion
        configurer.setUseTrailingSlashMatch(Optional.ofNullable(geoServer).map(g -> g.getGlobal()).map(g -> g.isTrailingSlashMatch()).orElse(true);
```

File: src/community/ogcapi/ogcapi-core/src/main/java/org/geoserver/ogcapi/APIDispatcher.java [hunk 4]
```diff
@@ -133,6 +146,8 @@ public UrlPathHelper getUrlPathHelper() {
         APIConfigurationSupport configurationSupport =
                 context.getBean(APIConfigurationSupport.class);
         configurationSupport.setCallbacks(callbacks);
+        PathMatchConfigurer configurer = new PathMatchConfigurer();
+        configurationSupport.configurePathMatch(configurer);
```
Comment (aaime) [inline src/community/ogcapi/ogcapi-core/src/main/java/org/geoserver/ogcapi/APIDispatcher.java:150]: Why is this needed?

File: src/web/core/src/main/resources/GeoServerApplication.properties [hunk 5]
```diff
@@ -348,6 +348,7 @@ GlobalSettingsPage.title             = Global Settings
 GlobalSettingsPage.verboseExceptions = Include stack trace in service exception
 GlobalSettingsPage.verboseMessaging  = Verbose XML output (pretty print) 
 GlobalSettingsPage.xmlExternalEntitiesEnabled = Unrestricted XML External Entity Resolution (security risk)
+GlobalSettingsPage.trailingSlashMatch = Match URLs with trailing slash (Requires restart)
```
Comment (aaime) [inline src/web/core/src/main/resources/GeoServerApplication.properties:351]: Nice that you added it to the label, as it's unusual. The screenshot is missing that bit though.

**Review by aaime — CHANGES_REQUESTED** [review]
Overall looks good, found just a few minor bits to comment on.
The default should be that we match trailing slash though, at least for now, we'll see what to do about it when we upgrade to Spring 6.

Changing all the tests so that they don't use a trailing match is probably a good idea, as they work with both settings, and will avoid having lots of overhead during an upgrade.
I'd leave just one, to make sure the current code can, by default, match trailing slashes.

Can we also have a test of flipping the behavior? GeoServer rebuilds the dispatchers at spring startup, so we need the test to be isolated from others.... the test should leverage data dir setup in order to change the config files before the dispatchers are built, I suppose, it can be done overriding setUpTestData in a GeoServerSystemTestSupport .

A possible alternative would be to have the default value (both in the field declaration, and in the setter) be controlled by a system variable (dynamic lookup, not a static final one) that you could flip on/off during the test, on a data dir that has the flag set to null.

File: src/main/src/main/java/org/geoserver/config/impl/GeoServerInfoImpl.java (hunk 2 above)
Comment (aaime) [reply src/main/src/main/java/org/geoserver/config/impl/GeoServerInfoImpl.java:53]: This could be controlled by a system variable, e.g. ``Boolean.getValue(System.getProperty("myProperty", "true))``

File: src/main/src/main/java/org/geoserver/config/impl/GeoServerInfoImpl.java [hunk 6]
```diff
@@ -281,6 +283,16 @@ public Boolean isXmlExternalEntitiesEnabled() {
...
     }
 
+    @Override
+    public void setTrailingSlashMatch(Boolean trailingSlashMatch) {
+        this.trailingSlashMatch = trailingSlashMatch;
+    }
+
+    @Override
+    public Boolean isTrailingSlashMatch() {
+        return trailingSlashMatch == null ? Boolean.TRUE : trailingSlashMatch;
```
Comment (aaime) [inline src/main/src/main/java/org/geoserver/config/impl/GeoServerInfoImpl.java:293]: And then have this Boolean.TRUE use ``Boolean.getValue(System.getProperty("myProperty", "true))`` as well (move that one-liner in a utility method)

File: src/main/src/main/java/org/geoserver/config/GeoServerInfo.java (hunk 1 above)
Comment (turingtestfail) [reply src/main/src/main/java/org/geoserver/config/GeoServerInfo.java]: fixed

File: src/main/src/main/java/org/geoserver/config/impl/GeoServerInfoImpl.java (hunk 2 above)
Comment (turingtestfail) [reply src/main/src/main/java/org/geoserver/config/impl/GeoServerInfoImpl.java]: changed

File: src/rest/src/main/java/org/geoserver/rest/RestConfiguration.java (hunk 3 above)
Comment (turingtestfail) [reply src/rest/src/main/java/org/geoserver/rest/RestConfiguration.java]: changed

File: src/community/ogcapi/ogcapi-core/src/main/java/org/geoserver/ogcapi/APIDispatcher.java (hunk 4 above)
Comment (turingtestfail) [reply src/community/ogcapi/ogcapi-core/src/main/java/org/geoserver/ogcapi/APIDispatcher.java]: removed

File: src/web/core/src/main/resources/GeoServerApplication.properties (hunk 5 above)
Comment (turingtestfail) [reply src/web/core/src/main/resources/GeoServerApplication.properties:351]: screenshot updated

File: src/main/src/main/java/org/geoserver/config/impl/GeoServerInfoImpl.java (hunk 6 above)
Comment (turingtestfail) [reply src/main/src/main/java/org/geoserver/config/impl/GeoServerInfoImpl.java]: changed

**Review by aaime — APPROVED** [review]
